draft-barnes-mls-appsync.md

```diff
@@ -60,171 +61,225 @@ communicate. Applications based on MLS can integrate their state into this
 
 # Conventions and Definitions
 
 {::boilerplate bcp14-tagged}
 
-# Application State Synchronization
+This document uses many terms from {{!RFC9420}}. Some of these have
+names which are similar or may be confusing: GroupContext, GroupContext
+extension, ExtensionType, and GroupContextExtensions proposal.
 
-This document defines a new AppSync proposal. AppSync is a Safe Extension as
-defined in {{Section 2 of !I-D.ietf-mls-extensions}}, of type
-`extension_external_proposal`.
+# GroupContext extensions modification
 
-The `application_states` extension allows the application to inject state
-objects into the MLS key schedule. Changes to this state can be made out of
-band, or using the AppSync proposal. Using the AppSync proposal ensures that
-members of the MLS group have received the relevant state changes before they
-are reflected in the group's `application_states`.
-
-> **NOTE:** This design exposes the high-level structure of the application state
-> to MLS.  An alternative design would be to have the application state be opaque
-> to MLS.  There is a trade-off between generality and the complexity of the API
-> between the MLS implementation and the application.  An opaque design would give
-> the application more freedom, but require the MLS stack to call out to the
-> application to get the updated state as part of Commit processing.  This design
-> allows the updates to happen within the MLS stack, so that no callback is
-> needed, at the cost of forcing the application state to fit a certain structure.
-> It also potentially can result in smaller state updates in large groups.
-
-The state for Each `applicationId` in the `application_states` needs to conform
-to one of four basic types: an ordered array, an unordered array, a map, or an
-irreducible blob. This allows the AppSync proposal to efficiently modify a large
-application state object.
-
-The content of the `application_states` extension and the `AppSync` proposal are
-structured as follows:
+This document defines a new GroupContextExtensionsDiff proposal. It
+is not a Safe Extension as defined in {{Section 2 of
+!I-D.ietf-mls-extensions}}, because it can modify any GroupContext
+extension, including those defined in {{!RFC9420}}. It is intended as a
+complete replacement for the GroupContextExtensions Proposal.
+The GroupContextExtensionsDiff proposal does not require an UpdatePath,
+and may be sent by an authorized external sender.
 
 ~~~ tls
 enum {
-    irreducible(0),
-    map(1),
-    unorderedList(2),
-    orderedArray(3),
+    remove(0),
+    add(1),
+    replace(2),
+    diff(3),
     (255)
-} StateType;
+} OperationType;
+
+uint8 DiffType;
 
+struct {
+    ExtensionType group_context_extension;
```

Comment:
```suggestion
    ExtensionType extension_type;
```

One potential difficulty here that you should probably just note for now: I'm pretty sure that there's nothing in MLS right now that says that there can only be one extension per `ExtensionType` in the GroupContext.  If there were duplicates, this would be ambiguous.  You could fix this with something dumb like (a) defining a `noop` OperationType and (b) saying that the n-th ExtensionDiff of a given type affects the n-th entry.  That has the benefit of degrading to the normal thing in the sane case.